torch_xla/experimental/custom_kernel_from_jax.py

```diff
@@ -0,0 +1,397 @@
+            global_block_q_dq, query.shape[2]),
+        block_kv_dq=None if global_use_fused_bwd_kernel else min(
+            global_block_kv_dq, query.shape[2]),
+        use_fused_bwd_kernel=global_use_fused_bwd_kernel,
+        q_layout=splash_attention_kernel.QKVLayout[global_q_layout],
+        k_layout=splash_attention_kernel.QKVLayout[global_k_layout],
+        v_layout=splash_attention_kernel.QKVLayout[global_v_layout],
+    )
+
+    mask = splash_attention_mask.CausalMask(
+        shape=(query.shape[2], query.shape[2]))
+
+    # Apply local masking if local sliding attention is enabled.
+    if config.AttentionType_LOCAL_SLIDING:
+      if config.SLIDE_WINDOW_SIZE is None:
+        raise ValueError(
+            "Sliding_window_size must be set if Local Sliding attention type")
+      mask &= splash_attention_mask.LocalMask(
+          shape=(query.shape[2], query.shape[2]),
+          window_size=(config.SLIDE_WINDOW_SIZE, config.SLIDE_WINDOW_SIZE),
+          offset=0,
+      )
+
+    # Create multi-head mask
+    multi_head_mask = splash_attention_mask.MultiHeadMask(
+        masks=(mask,) * query.shape[1])
+    splash_kernel = splash_attention_kernel.make_splash_mha(
+        mask=multi_head_mask,
+        head_shards=1,
+        q_seq_shards=1,
+        block_sizes=block_sizes,
+        attn_logits_soft_cap=attn_logits_soft_cap,
+    )
+    return jax.vmap(splash_kernel)(
+        query, key, value, segment_ids=decoder_segment_ids)
+
+  devices_in_data_fsdp = mesh.shape["data"] * mesh.shape["fsdp"]
+  assert (query.shape[0] / devices_in_data_fsdp).is_integer(), (
+      "Batch dimension should be shardable among the devices in data and fsdp axis"
+  )
+  x = wrap_flash_attention(query, key, value, decoder_segment_ids)
+  # x.shape = [batch, heads, seq_length, head_dim]
+  return x
+
+
+@functools.lru_cache(maxsize=16)
+def _get_jax_forward_function(config_json: str, attn_logits_soft_cap,
+                              has_segment_ids):
+  """Cached factory function to create JAX forward functions"""
+  config = SplashAttentionConfig.from_json(config_json)
+  if has_segment_ids:
+    return functools.partial(
+        splash_attention_jax_wrapper,
+        config=config,
+        attn_logits_soft_cap=attn_logits_soft_cap,
+    )
+  else:
+    return functools.partial(
+        splash_attention_jax_wrapper,
+        decoder_segment_ids=None,
+        config=config,
+        attn_logits_soft_cap=attn_logits_soft_cap,
+    )
+
+
+@functools.lru_cache(maxsize=16)
+def _get_jax_backward_function(config_json: str, attn_logits_soft_cap,
+                               has_segment_ids):
+  """Cached factory function to create JAX backward functions"""
+  jax_f = _get_jax_forward_function(config_json, attn_logits_soft_cap,
+                                    has_segment_ids)
+  import jax
+
+  if has_segment_ids:
+
+    def jax_grad_f_wrapper(query, key, value, decoder_segment_ids, grad_output):
```

Comment:
Similar to the forward case, I think this could be made into a global function that takes additional arguments such as the config object and the decoder_segment_ids as a kwargs, this avoiding the need for an additional caching layer.